File: 371_sum-of-two-integers.cpp
class Solution {
public:
    int getSum(int a, int b) {
        if (a < b) swap(a, b);
        int result = 0;
        for (int i = 1; i != 0; i <<= 1) {
            int ai = (a & i), bi = (b & i), carry = (result & i);
            if (ai & bi & carry) {
                result |= i;
                result |= (i << 1);
            } else if ((ai & bi) || (ai & carry) || (bi & carry)) {
                result &= ~(i);
                result |= (i << 1);
            } else if (ai || bi) result |= i;
        }
        return result;
    }
};

File: 371_sum-of-two-integers.py
class Solution(object):
    def getSum(self, a, b):
        """
        :type a: int
        :type b: int
        :rtype: int
        """
        MAX = 0x7FFFFFFF
        mask = 0xFFFFFFFF
        while b:
            a, b = (a ^ b) & mask, ((a & b) << 1) & mask
        return a if a <= MAX else ~(a ^ mask)
